### PR TITLE
Fix mutator footgun and types

### PR DESCRIPTION
**Pull request by jpaddison3**
The motivation for this PR was me wasting time because I assumed that data would be useful in an onCreate schemaField callback. Instead it's a footgun.

To explain more: `createMutator` takes in a new document to insert. Normally that document is take in via the `document` argument, and, when callbacks are accessing it, they can grab it from their arguments with `document`. However, long ago Vulcan used to support passing in the new document via an argument called `data`. They maintained backwards compatibility with it as an argument to createMutator, but did not ensure that data would get the information from document when passing it in as a `properties` value.

What follows is an attempt to:
1. Remove the createMutator data argument and property
2. Improve types well enough that we can be reasonably confident we've caught any usages, as well as because it's generally useful to have the mutators be typed when accessed from `Utils`

This is a mildly risky PR. I'm not confident it's net-positive. I'm open to hearing arguments that I should put `data` back in with deprecation warnings.

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1203853349753355) by [Unito](https://www.unito.io)

### CODE REVIEW

File: packages/lesswrong/lib/collections/subscriptions/mutations.ts [hunk 1]
```diff
@@ -21,8 +21,8 @@ export const defaultSubscriptionTypeTable = {
  */
 export const performSubscriptionAction = async (action, collection, itemId, user) => {
   const collectionName = collection.options.collectionName
-  const newSubscription = {
-    state: action === "subscribe" ? 'subscribed' : 'supressed',
+  const newSubscription: Partial<DbSubscription> = {
+    state: action === "subscribe" ? 'subscribed' : 'suppressed',
```
Comment (jpaddison3) [inline packages/lesswrong/lib/collections/subscriptions/mutations.ts:25]: Hey look a real bug. (Probably in unused code?)

File: packages/lesswrong/lib/types/collectionTypes.ts [hunk 2]
```diff
@@ -254,4 +254,46 @@ interface SpotlightFirstPost {
...
+  selector?: MongoSelector<T>,
+  data?: Partial<DbInsertion<T>>;
+  set?: Partial<DbInsertion<T>>;
+  unset?: any;
+  currentUser?: DbUser | null;
+  validate?: boolean;
+  context?: ResolverContext;
+  document?: T | null;
+};
+type UpdateMutator = <T extends DbObject>(args: UpdateMutatorParams<T>) => Promise<{ data: T }>;
```
Comment (Will-Howard) [inline packages/lesswrong/lib/types/collectionTypes.ts:285]: Slightly verbose but will this do the trick to fix the "selector or id" issue?
```suggestion
type UpdateMutatorParamsBase<T extends DbObject> = {
  collection: CollectionBase<T>;
  data?: Partial<DbInsertion<T>>;
  set?: Partial<DbInsertion<T>>;
  unset?: any;
  currentUser?: DbUser | null;
  validate?: boolean;
  context?: ResolverContext;
  document?: T | null;
};
type UpdateMutatorParamsWithDocId<T extends DbObject> = UpdateMutatorParamsBase<T> & {documentId: string, selector?: never};
type UpdateMutatorParamsWithSelector<T extends DbObject> = UpdateMutatorParamsBase<T> & {documentId?: never, selector: MongoSelector<T>};
type UpdateMutatorParams<T extends DbObject> = UpdateMutatorParamsWithDocId<T> | UpdateMutatorParamsWithSelector<T>;

type UpdateMutator = <T extends DbObject>(args: UpdateMutatorParams<T>) => Promise<{ data: T }>;
```
I got the `never` trick from [here](https://stackoverflow.com/questions/37688318/typescript-interface-possible-to-make-one-or-the-other-properties-required), I think you could also use undefined

Comment (Will-Howard) [reply packages/lesswrong/lib/types/collectionTypes.ts:285]: This does catch one place (introduced by me :( ) where both are used when only one is needed

**Review by Will-Howard — APPROVED** [review]
LGTM. I've tried reasonably hard to find places where `data` might be used implicitly in a way that would break this and I couldn't find any 👍

File: packages/lesswrong/lib/types/collectionTypes.ts (hunk 2 above)
Comment (jpaddison3) [reply packages/lesswrong/lib/types/collectionTypes.ts:285]: Yeah, maybe this is good. Maybe we should keep selector around tbh.

**Pull request merged**